src/detect-transform-base64.c

```diff
@@ -0,0 +1,381 @@
+#ifdef UNITTESTS
+static void DetectTransformFromBase64DecodeRegisterTests(void);
+#endif
+static void TransformFromBase64Decode(InspectionBuffer *buffer, void *options);
+
+#define DETECT_TRANSFORM_FROM_BASE64_MODE_DEFAULT (uint8_t) Base64ModeRFC4648
+
+void DetectTransformFromBase64DecodeRegister(void)
+{
+    sigmatch_table[DETECT_TRANSFORM_FROM_BASE64].name = "from_base64";
+    sigmatch_table[DETECT_TRANSFORM_FROM_BASE64].desc = "convert the base64 decode of the buffer";
+    sigmatch_table[DETECT_TRANSFORM_FROM_BASE64].url = "/rules/transforms.html#from_base64";
+    sigmatch_table[DETECT_TRANSFORM_FROM_BASE64].Setup = DetectTransformFromBase64DecodeSetup;
+    sigmatch_table[DETECT_TRANSFORM_FROM_BASE64].Transform = TransformFromBase64Decode;
+    sigmatch_table[DETECT_TRANSFORM_FROM_BASE64].Free = DetectTransformFromBase64DecodeFree;
+#ifdef UNITTESTS
+    sigmatch_table[DETECT_TRANSFORM_FROM_BASE64].RegisterTests =
+            DetectTransformFromBase64DecodeRegisterTests;
+#endif
+    sigmatch_table[DETECT_TRANSFORM_FROM_BASE64].flags |= SIGMATCH_OPTIONAL_OPT;
+}
+
+static void DetectTransformFromBase64DecodeFree(DetectEngineCtx *de_ctx, void *ptr)
+{
+    SCTransformBase64Free(ptr);
+}
+
+static SCDetectTransformFromBase64Data *DetectTransformFromBase64DecodeParse(const char *str)
+{
+    SCDetectTransformFromBase64Data *tbd = SCTransformBase64Parse(str);
+    if (tbd == NULL) {
+        SCLogError("invalid transform_base64 values");
+    }
+    return tbd;
+}
+
+/**
+ *  \internal
+ *  \brief Base64 decode the input buffer
+ *  \param det_ctx detection engine ctx
+ *  \param s signature
+ *  \param opts_str transform options, if any
+ *  \retval 0 No decode
+ *  \retval >0 Decoded byte count
+ */
+static int DetectTransformFromBase64DecodeSetup(
+        DetectEngineCtx *de_ctx, Signature *s, const char *opts_str)
+{
+    int r = -1;
+
+    SCEnter();
+
+    SCDetectTransformFromBase64Data *b64d = DetectTransformFromBase64DecodeParse(opts_str);
+    if (b64d == NULL)
+        SCReturnInt(r);
+
+    if (b64d->flags & DETECT_TRANSFORM_BASE64_FLAG_OFFSET_VAR) {
+        SCLogError("offset value must be a value, not a variable name");
+        goto exit_path;
+    }
+
+    if (b64d->flags & DETECT_TRANSFORM_BASE64_FLAG_NBYTES_VAR) {
+        SCLogError("byte value must be a value, not a variable name");
+        goto exit_path;
+    }
+
+    r = DetectSignatureAddTransform(s, DETECT_TRANSFORM_FROM_BASE64, b64d);
+
+exit_path:
+    if (r != 0)
+        DetectTransformFromBase64DecodeFree(de_ctx, b64d);
+    SCReturnInt(r);
+}
+
+static void TransformFromBase64Decode(InspectionBuffer *buffer, void *options)
+{
+    SCDetectTransformFromBase64Data *b64d = options;
+    const uint8_t *input = buffer->inspect;
+    const uint32_t input_len = buffer->inspect_len;
+    /*
+     * The output buffer is larger than needed. On average,
+     * a base64 encoded string is 75% of the encoded
```

Comment:
nit: decoded string is 75%

Reply:
Good point ... after the release occurs, I'll update this.